### PR TITLE
Fix bug in multi-panel plotting in bokeh

### LINKED ISSUE #1975: bokeh backend fails for some multi-panel plots
```
In [1]: from sherpa.astro import ui
WARNING: failed to import sherpa.astro.xspec; XSPEC models will not be available

In [2]: ui.set_plot_backend("BokehBackend")

In [3]: ui.load_pha("sherpa/astro/datastack/tests/data/3c273.pi")
WARNING: systematic errors were not found in file 'sherpa/astro/datastack/tests/data/3c273.pi'
statistical errors were found in file 'sherpa/astro/datastack/tests/data/3c273.pi'
but not used; to use them, re-read with use_errors=True
read ARF file sherpa/astro/datastack/tests/data/3c273.arf
read RMF file sherpa/astro/datastack/tests/data/3c273.rmf
WARNING: systematic errors were not found in file 'sherpa/astro/datastack/tests/data/3c273_bg.pi'
statistical errors were found in file 'sherpa/astro/datastack/tests/data/3c273_bg.pi'
but not used; to use them, re-read with use_errors=True
read background file sherpa/astro/datastack/tests/data/3c273_bg.pi

In [4]: ui.set_source(1, ui.powlaw1d(name='powlaw'))

In [5]: ui.fit()
WARNING: data set 1 has associated backgrounds, but they have not been subtracted, nor have background models been set
Dataset               = 1
Method                = levmar
Statistic             = chi2gehrels
Initial fit statistic = 8.15838e+10
Final fit statistic   = 50.5351 at function evaluation 25
Data points           = 46
Degrees of freedom    = 44
Probability [Q-value] = 0.231118
Reduced statistic     = 1.14852
Change in statistic   = 8.15838e+10
   powlaw.gamma   1.72329      +/- 0.0501892   
   powlaw.ampl    0.000147339  +/- 8.26966e-06 

In [6]: ui.plot_fit_delchi()
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
Cell In[6], line 1
----> 1 ui.plot_fit_delchi()

File <string>:1, in plot_fit_delchi(id, replot, overplot, clearwindow, **kwargs)

File ~/code/sherpa/sherpa/ui/utils.py:14799, in Session.plot_fit_delchi(self, id, replot, overplot, clearwindow, **kwargs)
  14797 plot1obj = self.get_fit_plot(id, recalc=not replot)
  14798 plot2obj = self.get_delchi_plot(id, recalc=not replot)
> 14799 self._jointplot2(plot1obj, plot2obj,
  14800                  overplot=overplot, clearwindow=clearwindow,
  14801                  **kwargs)

File ~/code/sherpa/sherpa/ui/utils.py:14509, in Session._jointplot2(self, plot1, plot2, overplot, clearwindow, **kwargs)
  14503 self._jointplot.reset()
  14505 with sherpa.plot.backend:
  14506 
  14507     # Note: the user preferences are set to both plots
  14508     #
> 14509     self._jointplot.plottop(plot1, overplot=overplot,
  14510                             clearwindow=clearwindow, **kwargs)
  14512     # The two plots are intended to have the same scaling
  14513     # on the X axis (log or linear), and the approach is
  14514     # to use log if either of the components of the top
   (...)
  14520     # types of plot objects.
  14521     #
  14522     oldval = plot2.plot_prefs['xlog']

File ~/code/sherpa/sherpa/plot/__init__.py:1260, in JointPlot.plottop(self, plot, overplot, clearwindow, *args, **kwargs)
   1257 if create:
   1258     self._clear_window()
-> 1260 backend.set_jointplot(0, 0, self.rows, self.cols,
   1261                       create=create)
   1263 # The code used to check if the plot was an instance of
   1264 # FitPlot, which has been updated to check for the presence
   1265 # of attributes instead.
   1266 #
   1267 if hasattr(plot, 'xlabel'):

File ~/code/sherpa/sherpa/plot/bokeh_backend.py:743, in BokehBackend.set_jointplot(self, row, col, nrows, ncols, create, top, ratio)
    741 for fig in enumerate(self.current_fig.children):
    742     if fig[1] != top:
--> 743         height = fig[0].height
    744         break
    745 # normal_hight is None if we only have one row, so we don't need to
    746 # change anything.

AttributeError: 'int' object has no attribute 'height'
```

I don't know why I did not discover that in testing, but I guess that's why we labelled this backend "experimental".

### PR DESCRIPTION
**Pull request by hamogu**
## Summary
Fix bug in multi-panel plotting in bokeh

## Details

There are in fact two separate problems:

- The current figure sis a GridPlot and no longer (as originally envisioned by me) a plain Python list, so the indexing needs to index the children of the GridPlot not a plain list.
- Color values need to be translated.

Both should have been caught in the original bokeh backend implementation (and I'm sure the color translation was there at some point, but probably got lost in one of the many rebases I went through). I've added just one test that checks both problems.

closes #1975

### CODE REVIEW

File: sherpa/plot/tests/test_bokeh.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+        "tests",
+        "data",
+        "3c273.pi",
+    )
+    ui.load_pha(path)
+    ui.set_source(1, ui.powlaw1d(name="powlaw"))
+
+    bokehbackend = pytest.importorskip("sherpa.plot.bokeh_backend")
+    newback = bokehbackend.BokehBackend()
+    with plot.TemporaryPlottingBackend(newback):
```
Comment (DougBurke) [inline sherpa/plot/tests/test_bokeh.py:54]: If we add any more tests we probably want to add a fixture to `conftest.py` - `requires_bokeh` to complement the new `requires_pylab` fixture.

Comment (hamogu) [reply sherpa/plot/tests/test_bokeh.py:54]: If and when we add more tests. One thing that's beyond the scope of this PR is to actually test the output. bokeh plots get turned into json (which is then passed to the browser), so we can do tests that check the text of plot labels like `assert "counts / s / keV" in outputplot_json`. Those things are hard to test properly in matplotlib! I envision that we add tests like that over time and then we'll see if we need a global or local fixture for this.

Comment (DougBurke) [reply sherpa/plot/tests/test_bokeh.py:54]: I don't think any of our CI runs include bokeh do they? Or did we change that when you added #1899 - I have a feeling that you did that, but am too lazy to look myself

Comment (hamogu) [reply sherpa/plot/tests/test_bokeh.py:54]: https://github.com/sherpa/sherpa/commit/87090c67bd42fce4288cbeba9465b02bd5cb3573

File: sherpa/plot/tests/test_bokeh.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+from sherpa.astro import ui
+from sherpa.utils.testing import requires_data, requires_fits
+
+
+@requires_fits
+@requires_data
+def test_bokeh_delchi(caplog, clean_ui):
+    """Test a specific multi-panel plot with the bokeh backend.
+
+    This also implictly test setting a line_color, because
```
Comment (wmclaugh) [inline sherpa/plot/tests/test_bokeh.py:34]:
```suggestion
    This also implicitly tests setting a line_color, because
```

**Comment by DougBurke** [comment]
SO, in checking https://github.com/sherpa/sherpa/commit/87090c67bd42fce4288cbeba9465b02bd5cb3573 I note that this adds bokeh to the "no astropy or xspec" tests (and they also do not have the test data present) which means that the `test_bokeh.py` test added here is not actually run in any of our CI runs. If you look at one of the test runs you can see it is skipped - e.g. 

https://github.com/sherpa/sherpa/actions/runs/8122816366/job/22202638267?pr=1976


Maybe we should add bokeh to at least one of our conda everything-and-the-kitchen-sink tests (but not the macOS one)?

**Comment by DougBurke** [comment]
@hamogu - do you have any thoughts on https://github.com/sherpa/sherpa/pull/1976#issuecomment-1977012956 ?

**Comment by hamogu** [comment]
> Maybe we should add bokeh to at least one of our conda everything-and-the-kitchen-sink tests (but not the macOS one)?

Sure, why not? It depends a bit on how much longer we treat bokeh as "experimental" and how hard it is. If we just add a test and it all works, fine. It it requires a major rewriting of our test setup, then I'd argue that it's not absolutely necessary. The tests in question still get run locally if you have bokeh and the test data, just not in CI. For an "experimental" pacakge that might be OK, so this is somewhat tied to the future of Sherpa: If we get to expand bokeh (as I hope) we should do that. If we need to trim Sherpa down to the absolute core, we might not need it.

**Comment by hamogu** [comment]
Do we want to do changes in CI as part of this PR? Oe treat that separately?

**Comment by DougBurke** [comment]
I have two commits - one to `sherpa/astro/ui/tests/test_astro_ui_plot.py` that shows the resid issue in #1991 is addressed and adds a CI run with bokeh. I could add them to this PR or you can look at

# CI

```diff
diff --git a/.github/scripts/setup_conda.sh b/.github/scripts/setup_conda.sh
index 3312e85d..71cc5674 100755
--- a/.github/scripts/setup_conda.sh
+++ b/.github/scripts/setup_conda.sh
@@ -16,7 +16,7 @@ else
 
     if [ -n "${MATPLOTLIBVER}" ]; then
         #Installed for qt-main deps which is needed for the QtAgg backend to work for matplotlib
-        #This is only an issue on stripped down systems. You can check for this issue by: 
+        #This is only an issue on stripped down systems. You can check for this issue by:
         #  ldd $CONDA_PREFIX/plugins/platforms/libqxcb.so | grep "not found"
         sudo apt-get install -q libxcb-icccm4 libxcb-image0 libxcb-keysyms1 libxcb-render-util0 libxcb-icccm4 libxcb-image0 libxcb-keysyms1 libxcb-render-util0
    fi
@@ -42,16 +42,17 @@ fi
 
 # Figure out requested dependencies
 if [ -n "${MATPLOTLIBVER}" ]; then MATPLOTLIB="matplotlib=${MATPLOTLIBVER}"; fi
+if [ -n "${BOKEHVER}" ]; then BOKEH="bokeh=${BOKEHVER}"; fi
 if [ -n "${NUMPYVER}" ]; then NUMPY="numpy=${NUMPYVER}"; fi
 # Xspec >=12.10.1n Conda package includes wcslib & CCfits and pulls in cfitsio & fftw
 if [ -n "${XSPECVER}" ];
  then export XSPEC="xspec-modelsonly=${XSPECVER}";
 fi
 
-echo "dependencies: ${MATPLOTLIB} ${NUMPY} ${FITS} ${XSPEC}"
+echo "dependencies: ${MATPLOTLIB} ${BOKEH} ${NUMPY} ${FITS} ${XSPEC}"
 
 # Create and activate conda build environment
-conda create --yes -n build python"=${PYTHONVER}.*=*cpython*" pip ${MATPLOTLIB} ${NUMPY} ${XSPEC} ${FITSBUILD} ${compilers}
+conda create --yes -n build python"=${PYTHONVER}.*=*cpython*" pip ${MATPLOTLIB} ${BOKEH} ${NUMPY} ${XSPEC} ${FITSBUILD} ${compilers}
 
 conda activate build
 
diff --git a/.github/workflows/ci-conda-workflow.yml b/.github/workflows/ci-conda-workflow.yml
index 0ea39471..e1e0d8d9 100644
--- a/.github/workflows/ci-conda-workflow.yml
+++ b/.github/workflows/ci-conda-workflow.yml
@@ -53,6 +53,7 @@ jobs:
             fits: astropy
             test-data: submodule
             matplotlib-version: 3
+            bokeh-version: 3
             xspec-version: 12.13.1
 
           - name: Linux Full Build (Python 3.10)
@@ -106,6 +107,7 @@ jobs:
         NUMPYVER: ${{ matrix.numpy-version }}
         FITSBUILD: ${{ matrix.fits }}
         MATPLOTLIBVER: ${{ matrix.matplotlib-version }}
+        BOKEHVER: ${{ matrix.bokeh-version }}
         XSPECVER: ${{ matrix.xspec-version }}
       run: |
         source .github/scripts/setup_conda.sh
```

# Test

```diff
diff --git a/sherpa/astro/ui/tests/test_astro_ui_plot.py b/sherpa/astro/ui/tests/test_astro_ui_plot.py
index b3b26ed4..eb0ac117 100644
--- a/sherpa/astro/ui/tests/test_astro_ui_plot.py
+++ b/sherpa/astro/ui/tests/test_astro_ui_plot.py
@@ -1099,7 +1099,7 @@ def check_bkg_source(plotfunc, idval, isfit=True):
                           (ui.plot_bkg_fit_ratio, [check_bkg_fit, check_bkg_resid]),
                           (ui.plot_bkg_fit_resid, [check_bkg_fit, check_bkg_resid])])
 def test_bkg_plot_xxx(idval, plotfunc, checkfuncs,
-                      clean_astro_ui, requires_pylab):
+                      clean_astro_ui, all_plot_backends_astro_ui):
     """Test background plotting - channel space"""
 
     setup_example_bkg_model(idval)
@@ -1179,7 +1179,7 @@ _basic_plotfuncs = [ui.plot_data,
 
 @requires_fits
 @requires_data
-def test_pha1_plot_function(clean_astro_ui, basic_pha1):
+def test_pha1_plot_function(clean_astro_ui, basic_pha1, all_plot_backends_astro_ui):
     # can we call plot; do not try to be exhaustive
     ui.plot("data", "bkg", "fit", "arf")
 
@@ -1187,7 +1187,7 @@ def test_pha1_plot_function(clean_astro_ui, basic_pha1):
 @requires_fits
 @requires_data
 @pytest.mark.parametrize("plotfunc", _basic_plotfuncs)
-def test_pha1_plot(plotfunc, clean_astro_ui, basic_pha1):
+def test_pha1_plot(plotfunc, clean_astro_ui, basic_pha1, all_plot_backends_astro_ui):
     plotfunc()
 
 
@@ -1264,7 +1264,7 @@ def test_pha1_plot_clearwindow(hide_logging, clean_astro_ui,
                                       ui.plot_bkg_fit_ratio,
                                       ui.plot_bkg_fit_delchi])
 def test_pha1_bkg_plot(plotfunc, clean_astro_ui, basic_pha1,
-                       hide_logging):
+                       hide_logging, all_plot_backends_astro_ui):
     """Test issue #943 and general check of plot_bkg_xxx"""
 
     ui.unsubtract()
@@ -1276,7 +1276,7 @@ def test_pha1_bkg_plot(plotfunc, clean_astro_ui, basic_pha1,
 @requires_data
 @pytest.mark.parametrize("plotfunc", [ui.plot_model_component,
                                       ui.plot_source_component])
-def test_pha1_plot_component(clean_astro_ui, basic_pha1, plotfunc):
+def test_pha1_plot_component(clean_astro_ui, basic_pha1, plotfunc, all_plot_backends_astro_ui):
     plotfunc("pl")
 
 
@@ -1522,7 +1522,7 @@ def test_pha1_plot_with_model_component_no_response(clean_astro_ui,
 
 @requires_fits
 @requires_data
-def test_pha1_plot_order(clean_astro_ui, basic_pha1):
+def test_pha1_plot_order(clean_astro_ui, basic_pha1, all_plot_backends_astro_ui):
     ui.plot_order()
 
 
@@ -3444,7 +3444,7 @@ def test_datapha_plot_after_clean():
                           'resid',
                           'ratio',
                           'delchi'])
-def test_set_plot_opt_x(cls, datafunc, plotfunc):
+def test_set_plot_opt_x(cls, datafunc, plotfunc, all_plot_backends):
     """Does set_xlog/xlinear work?
 
     We run for both session types here, rather than duplicate
@@ -4393,7 +4393,7 @@ def test_pha_model_plot_filter_range_1024_false(mask, expected, make_data_path,
 @requires_data
 @requires_wcs
 @pytest.mark.parametrize("coord", ["logical", "image", "physical", "world", "wcs"])
-def test_1380_plot(coord, make_data_path, clean_astro_ui):
+def test_1380_plot(coord, make_data_path, clean_astro_ui, all_plot_backends_astro_ui):
     """The contour data should ideally remain the same.
 
     See also sherpa/astro/tests/test_astro_data2.py::test_1380_data
```

**Comment by DougBurke** [comment]
I've just pushed the commit that adds bokeh to one of the CI runs. When that passes (as there's no chance of it failing) I'll push the test commit.

**Comment by DougBurke** [comment]
Look at that bump in coverage!

**Comment by DougBurke** [comment]
@hamogu - can you just check my last two commits don't have anything obviously wrong in them?

**Comment by hamogu** [comment]
LGTM

**Review by DougBurke — APPROVED** [review]
Now we've added some tests that exercise some of the problem areas and run tests with bokeh (although slightly less important than my original concern now we've improved the test suite) I'm happy with this.

**Pull request merged**